lib/communicator_stack.cpp

```diff
@@ -0,0 +1,215 @@
+#include <communicator_quda.h>
+#include <map>
+#include <array>
+#include <lattice_field.h>
+
+int Communicator::gpuid = -1;
+
+static std::map<quda::CommKey, Communicator> communicator_stack;
+
+static quda::CommKey current_key = {-1, -1, -1, -1};
+
+void init_communicator_stack(int ndim, const int *dims, QudaCommsMap rank_from_coords, void *map_data,
+                             bool user_set_comm_handle, void *user_comm)
+{
+  communicator_stack.emplace(
+    std::piecewise_construct, std::forward_as_tuple(default_comm_key),
+    std::forward_as_tuple(ndim, dims, rank_from_coords, map_data, user_set_comm_handle, user_comm));
+
+  current_key = default_comm_key;
+}
+
+void finalize_communicator_stack() { communicator_stack.clear(); }
+
+static Communicator &get_default_communicator()
+{
+  auto search = communicator_stack.find(default_comm_key);
+  if (search == communicator_stack.end()) { errorQuda("Default communicator can't be found."); }
+  return search->second;
+}
+
+Communicator &get_current_communicator()
+{
+  auto search = communicator_stack.find(current_key);
+  if (search == communicator_stack.end()) { errorQuda("Current communicator can't be found."); }
+  return search->second;
+}
+
+void push_communicator(const quda::CommKey &split_key)
+{
+  auto search = communicator_stack.find(split_key);
+  if (search == communicator_stack.end()) {
+    communicator_stack.emplace(std::piecewise_construct, std::forward_as_tuple(split_key),
+                               std::forward_as_tuple(get_default_communicator(), split_key.data()));
+  }
+
+  quda::LatticeField::freeGhostBuffer(); // Destroy the (IPC) Comm buffers with the old communicator.
+
+  current_key = split_key;
+}
+
+int comm_neighbor_rank(int dir, int dim) { return get_default_communicator().comm_neighbor_rank(dir, dim); }
```

Comment:
This `get_default_communicator().`... seems messy... but I'm guessing this is the path of least resistance approach?

Reply:
It's a bug. Great you caught it. `comm_neighbor_rank` is never used and this is never exposed.

Reply:
Haha, excellent unintended consequence. Is this true for any of the other routines?